--- a/lib/parse.js
+++ b/lib/parse.js
@@ -1,6 +1,7 @@
 'use strict';
 
 const stringify = require('./stringify');
+const utils = require('./utils');
 
 /**
  * Constants
@@ -143,6 +144,7 @@ const parse = (input, options = {}) => {
      */
 
     if (value === CHAR_LEFT_PARENTHESES) {
+      utils.assertDepth(stack.length);
       block = push({ type: 'paren', nodes: [] });
       stack.push(block);
       push({ type: 'text', value });
@@ -203,6 +205,7 @@ const parse = (input, options = {}) => {
      */
 
     if (value === CHAR_LEFT_CURLY_BRACE) {
+      utils.assertDepth(stack.length);
       depth++;
 
       const dollar = prev.value && prev.value.slice(-1) === '$' || block.dollar === true;
